lang/image.bzl

```diff
@@ -34,8 +34,8 @@ def _binary_name(ctx):
     # /app/foo/bar/baz/blah
     return "/".join([
         ctx.attr.directory,
-        ctx.attr.binary.label.package,
-        ctx.attr.binary.label.name,
+        ctx.attr.binary[0].label.package,
```

Comment:
for this to be safe, it seems like the `binary` attribute would have to be declared with `allow_single_file=true`? how do you know the zeroth element should be used?

Reply:
I also feel it's an odd design decision, but it is documented behaviour when using an [outgoing edge transition](https://docs.bazel.build/versions/main/skylark/config.html#accessing-attributes-with-transitions).

> When attaching a transition to an outgoing edge (regardless of whether the transition is a 1:1 or 1:2+ transition) ctx.attr is forced to be a list if it isn’t already. The order of elements in this list is unspecified.

It looks like we aren't the only ones who looked twice as per this comment in [rules_go](https://github.com/bazelbuild/rules_go/blob/aaf48cbd40dbc74a6199a1ff18db72090b11952b/go/private/rules/transition.bzl#L266).

I would imagine it's for [1:2+ transitions](https://docs.bazel.build/versions/main/skylark/config.html#defining-12-transitions):

> Outgoing edge transitions can map a single input configuration to two or more output configurations.

In such scenarios, there may be multiple binaries. However for our 1:1 transition, it will only ever have a single item.

Reply:
Assuming we're going to use an incoming edge transition instead, this may no longer be relevant.